File: 2022/Day 2/pt1.cpp
#include <helper.hpp>
#include <map>

// standardise using just "A, B, C"
std::map<char, char> decoder_map{{'X', 'A'}, {'Y', 'B'}, {'Z', 'C'}};
std::map<char, int> score_map{{'A', 1}, {'B', 2}, {'C', 3}};
std::map<char, std::map<char, int>> round_score_map;

void populate_round_score_map()
{
    round_score_map['A'] = {{'A', 3}, {'B', 6}, {'C', 0}};
    round_score_map['B'] = {{'A', 0}, {'B', 3}, {'C', 6}};
    round_score_map['C'] = {{'A', 6}, {'B', 0}, {'C', 3}};
}

int get_total_score(std::vector<std::string> lines)
{
    int total_score = 0;

    for (std::string line : lines)
    {
        char opnt, your;
        sscanf(line.c_str(), "%c %c", &opnt, &your);
        your = decoder_map[your];

        total_score += round_score_map[opnt][your] + score_map[your];
    }

    return total_score;
}

int main()
{
    std::vector<std::string> lines = Helper::get_lines("input.txt");

    auto t1 = Helper::get_current_time();

    populate_round_score_map();
    int score = get_total_score(lines);
    std::cout << "Output: " << score << std::endl;

    auto t2 = Helper::get_current_time();
    Helper::get_time_taken(t1, t2);

    return 0;
}

File: 2023/day_02/pt1.go
package day02

import (
	"fmt"
	"regexp"
	"strconv"
	"strings"

	"github.com/benjababe/advent-of-code/helper"
)

func solvePt1(lines []string) int {
	score := 0
	cubeMax := map[string]int{"r": 12, "g": 13, "b": 14}

	for i, line := range lines {
		line = line[strings.Index(line, ":")+1:]
		possible := true

		r := regexp.MustCompile(`(\d+)\s(r|g|b)`)
		matches := r.FindAllStringSubmatch(line, -1)

		for _, match := range matches {
			count, _ := strconv.Atoi(match[1])
			if count > cubeMax[match[2]] {
				possible = false
				break
			}
		}

		if possible {
			score += i
		}
	}

	return score
}

func Pt1() {
	lines := []string{}
	helper.GetLines(&lines, "input.txt")

	start := helper.GetCurrentTime()
	output := solvePt1(lines)
	fmt.Printf("Day 2\tPt1:\t%d\n", output)
	end := helper.GetCurrentTime()
	helper.GetTimeTaken(start, end)
}
